### PR TITLE
Returns nil when vim.fn is indexed with invalid key. Fixes #19271.

### LINKED ISSUE #19271: Lua: vim.fn['bogus'] return nil for nonexistent function
### Feature already in Vim?

No

### Feature description

Lua config files have the ability to call functions in Vimscript, e.g. `vim.fn.expand("%:p")`.

However, there is no way to check if a function exists before trying to call it.

One use case for that is a config file which degrades gracefully if certain features are not available. For example, if `vim-plug` is installed, setup certain things, otherwise do other things. We could check if the function `vim.fn['plug#']` exists or not.

Currently a workaround is to just attempt to call it, and handle the failure...

```lua
local success, error = pcall(vim.fn['plug#begin'], vim.fn.stdpath('data') .. '/plugged')
if not success then --[[fail gracefully]] end
```

That works for this use case, but might have issues if the function has side effects and is not idempotent. IMHO, it would be cleaner to have a conditional like `if vim.fnexists("plug#") then ...`.

See [this SO post](https://stackoverflow.com/questions/72850062/neovim-check-if-a-vim-function-exists-from-lua/72850145) for original discussion.

### PR DESCRIPTION
**Pull request by addcninblue**

### CODE REVIEW

**Comment by ii14** [comment]
I don't think this behavior could be compatible with the current `vim.fn` implementation. If the function was first cached inside `vim.fn` but then deleted, `vim.fn` will throw errors from the cached function instead of returning nil.

File: runtime/lua/vim/_editor.lua [hunk 1]
```diff
@@ -272,6 +272,8 @@ vim.fn = setmetatable({}, {
       _fn = function()
         error(string.format('Tried to call API function with vim.fn: use vim.api.%s instead', key))
       end
+    elseif vim.call("exists", "*" .. key) == 0 then
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:275]: should check `t[key]` first, no need to check this if we already stored a value there.

Comment (justinmk) [inline runtime/lua/vim/_editor.lua:275]: should check `t[key]` first, no need to check this if we already stored a value there.

```suggestion
    elseif t[key] == nil and vim.call("exists", "*" .. key) == 0 then
```

File: test/functional/lua/vim_spec.lua [hunk 2]
```diff
@@ -795,6 +795,10 @@ describe('lua stdlib', function()
           pcall_err(exec_lua, "vim.fn.nvim_get_current_line()"))
   end)
 
+  it('vim.fn should return nil when accessing invalid function', function()
+    eq(true, exec_lua("vim.fn['bogus'] == nil"))
```
Comment (justinmk) [inline test/functional/lua/vim_spec.lua:799]: need a test for user-defined function , then also after it's deleted

**Comment by justinmk** [comment]
Thanks for exploring this. Decided against in https://github.com/neovim/neovim/issues/19271

File: runtime/lua/vim/_editor.lua (hunk 1 above)
Comment (ii14) [reply runtime/lua/vim/_editor.lua:275]: I think `t[key]` would call this `__index` metamethod and end up in a infinite recursive loop. `__index` is only called when the key does not exist on a table.

**Pull request closed without merging**